### PR TITLE
boards: arm: nucleo_f746zg: Created a new stm32 nucleo board.

### PR DESCRIPTION
**Pull request by ajpcode**
Added board support for the nucleo_f746zg.
Features supported include:
- GPIO
- UART
- ETH
- USB
- I2C
- PWM
- SPI

Signed-off-by: AJ Palmer <ajpcode@hotmail.com>

### CODE REVIEW

**Comment by therealprof** [comment]
@ajpcode Any reason why you're only bringing in a small subset of the peripherals? Ethernet works just fine on this board and would allow to use many of the networking samples as well. I do have a complete set of patches (sans the documentation, hence I haven't done a PR yet) and I'd be happy to share them with you if you'd like to work them in.

**Comment by ajpcode** [comment]
@therealprof Hi, this is my first attempt at a contribution so I'm a bit new to the process (didn't want to commit too much on my first attempt). I've also had Ethernet, SPI and I2C working albeit on v1.10 but if I can figure out how this contribution thing works I'd be happy to try and integrate your more up-to-date work.

**Comment by therealprof** [comment]
@ajpcode It's okay then, I don't want to overload you here. Let's hear back from the friendly Zephyr folks about your contribution and follow up with the additional stuff later.

**Review by erwango — CHANGES_REQUESTED** [review]
@ajpcode , thanks for this contribution.
As mentionned by @therealprof , let's provide full-fledged version of the board.
One thing though, please have a look to recent versions of nucleo_144 board (nucleo_f429zi for instance), and provide similar configuration.

**Comment by ajpcode** [comment]
@erwango - Thank you for your feedback. 

There isn't an awful lot of peripheral support for the F7 in Zephyr at the moment (ref: soc/arm/st_stm32/stm32f7/soc.h and boards/arm/stm32f746g_disco/doc/stm32f746g_disco.rst) - I was hoping to add some of it once I had a board to work from. 

I could maybe add I2C pinmux (sample/sensor/bme280 test) and maybe add GPIO definitions for LEDs/user buttons on the board (and test with samples/basic/blinky)?

I feel the device driver addition is probably worthy of it's own PR - at which point I can add the respective feature to the board. I suppose I'm trying not to over-extend this request into too many areas - how would you feel about this?

Kind regards.

**Comment by therealprof** [comment]
@ajpcode I've pushed my current draft version here: https://github.com/therealprof/zephyr/tree/preliminary-nucleo-f746zg-support

Feel free to pick up whatever pieces you see fit. I've been testing networking with this extensively, which is one of the key features of this board.

**Comment by ajpcode** [comment]
@therealprof I'm just having a look now. Ethernet seems a good place to start then I'll pull the pinmux and DT definitions for the LEDs and PB. Were you testing anything specific with the PWM (ref: pinmux.c)?

**Comment by therealprof** [comment]
Nope, haven't used PWM on this board. Most of the peripherals are equivalent to the similarly sized Nucleo-F429ZI, though, so if it works there chances are good it'll work here, too.

**Comment by ajpcode** [comment]
> Nope, haven't used PWM on this board. Most of the peripherals are equivalent to the similarly sized Nucleo-F429ZI, though, so if it works there chances are good it'll work here, too.

I agree that it will likely work (although I am unable to test right now) but as PWM appears to be broken out to CN7 (PD9, PD8) and is not hardwired to a device on the board it seems like something a user would specify for their individual projects rather than something the board might specify for them - that would be my interpretation although I am open to alternative thoughts.

Can you think of any other changes from your preliminary work that I've overlooked @therealprof ?

File: boards/arm/nucleo_f746zg/doc/nucleof746zg.rst [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+
+ST Nucleo F746ZG
+################
+
+Overview
+********
+
+The STM32 Nucleo-144 boards provide an affordable and flexible way for users to try out new concepts and build prototypes by choosing from the various combinations of performance and power consumption features, provided by the STM32 microcontroller. For the compatible boards, the SMPS significantly reduces power consumption in Run mode.
+
+The ST Zio connector, which extends the Arduino™ Uno V3 connectivity, and the ST morpho headers provide an easy means of expanding the functionality of the Nucleo open development platform with a wide choice of specialized shields.
```
Comment (dbkinder) [inline boards/arm/nucleo_f746zg/doc/nucleof746zg.rst]: Change to:

    The Arduino-compatible ST Zio connector expands functionality
    of the Nucleo open development platform, with a wide choice
    of specialized Arduino* Uno V3 shields.

File: boards/arm/nucleo_f746zg/doc/nucleof746zg.rst [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+
+The STM32 Nucleo-144 board does not require any separate probe as it integrates the ST-LINK/V2-1 debugger/programmer.
+
+The STM32 Nucleo-144 board comes with the STM32 comprehensive free software libraries and examples available with the STM32Cube MCU Package.
```
Comment (dbkinder) [inline boards/arm/nucleo_f746zg/doc/nucleof746zg.rst]: Change to:

    The STM32 Nucleo-144 board comes with comprehensive free software
    libraries and examples, available with the STM32Cube MCU Package.

File: boards/arm/nucleo_f746zg/doc/nucleof746zg.rst [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+
+Key Features
+
+    STM32 microcontroller in LQFP144 package
```
Comment (dbkinder) [inline boards/arm/nucleo_f746zg/doc/nucleof746zg.rst]: All these list items need a leading ``-``:

    - STM32 microcontroller in LQFP144 package
    - External SMPS ...

File: boards/arm/nucleo_f746zg/doc/nucleof746zg.rst [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+    -   USB with Micro-AB
+    -   SWD
+    -   Ethernet RJ45 (depending on STM32 support)
+    -   ST Zio connector including Arduino™ Uno V3
+    -   ST morpho
+
+    Flexible power-supply options: ST-LINK USB VBUS or external sources
+    On-board ST-LINK/V2-1 debugger/programmer with USB re-enumeration capability: mass storage, virtual COM port and debug port
+    Comprehensive free software libraries and examples available with the STM32Cube package
+    Support of a wide choice of Integrated Development Environments (IDEs) including IAR™, Keil®, GCC-based IDEs, Arm® Mbed™
```
Comment (dbkinder) [inline boards/arm/nucleo_f746zg/doc/nucleof746zg.rst]: We recognize trademarks of other companies by using an asterisk ``*`` and not a TM or (r) symbol.

File: boards/arm/nucleo_f746zg/doc/nucleof746zg.rst [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+
+The default configuration can be found in the defconfig file:
+``boards/arm/nucleo_f746zg/nucleo_f746zg_defconfig``
+
+For mode details please refer to `STM32 Nucleo-144 board User Manual`_.
+
+Default Zephyr Peripheral Mapping:
+----------------------------------
+
+TBC
```
Comment (dbkinder) [inline boards/arm/nucleo_f746zg/doc/nucleof746zg.rst]: Will this section be provided before asking this PR to be merged?

File: boards/arm/nucleo_f746zg/doc/nucleof746zg.rst [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+
+Nucleo F746ZG board has 4 UARTs and 4 USARTs. The Zephyr console output is assigned to UART2.
+Default settings are 115200 8N1.
+
+
+Programming and Debugging
+*************************
+
+Applications for the ``nucleo_f746zg`` board configuration can be built and
+flashed in the usual way (see `build_an_application` and
```
Comment (dbkinder) [inline boards/arm/nucleo_f746zg/doc/nucleof746zg.rst]: change to ``:ref:`build_and_application` ``

File: boards/arm/nucleo_f746zg/doc/nucleof746zg.rst [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+`application_run` for more details).
```
Comment (dbkinder) [inline boards/arm/nucleo_f746zg/doc/nucleof746zg.rst]: change to ``:ref:`application_run` ``

File: boards/arm/nucleo_f746zg/doc/nucleof746zg.rst [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+
+Flashing
+========
+
+Nucleo F746ZG board includes an ST-LINK/V2-1 embedded debug tool interface.
+
+Flashing an application to Nucleo F746ZG
+----------------------------------------
+
+Here is an example for the `hello_world` application.
```
Comment (dbkinder) [inline boards/arm/nucleo_f746zg/doc/nucleof746zg.rst]: change to ``:ref:`hello_world` ``

File: boards/arm/nucleo_f746zg/doc/nucleof746zg.rst [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+
+.. code-block:: console
+
+   $ Hello World! nucleo_f746zg
+
+Debugging
+=========
+
+You can debug an application in the usual way.  Here is an example for the
+`hello_world` application.
```
Comment (dbkinder) [inline boards/arm/nucleo_f746zg/doc/nucleof746zg.rst]: change to ``:ref:`hello_world` ``

File: boards/arm/nucleo_f746zg/doc/nucleof746zg.rst [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+   :goals: debug
+
+.. _Nucleo F746ZG website:
+   https://www.st.com/en/evaluation-tools/nucleo-f746zg.html
+
+.. _STM32 Nucleo-144 board User Manual:
+   http://www.st.com/resource/en/user_manual/dm00105823.pdf
+
+.. _STM32F746ZG on www.st.com:
+   https://www.st.com/content/st_com/en/products/microcontrollers/stm32-32-bit-arm-cortex-mcus/stm32-high-performance-mcus/stm32f7-series/stm32f7x6/stm32f746zg.html
```
Comment (dbkinder) [inline boards/arm/nucleo_f746zg/doc/nucleof746zg.rst]: this is a case where a long line length is OK.  Otherwise, please keep line length < 80

**Review by dbkinder — CHANGES_REQUESTED** [review]

**Comment by ajpcode** [comment]
@erwango - Would this be an acceptable initial feature set for this board (pinmux, gpio, uart, i2c, ethernet, usb + led and sw definitions) ?

**Review by erwango — CHANGES_REQUESTED** [review]
Overall looks good, missing few bits here and there.
Mainly some clean up to group changes into logical groups.

My advice is to squash all commits and then split into 3 commits:
-Board intro with UART, led and button + main doc
-Eth changes (dts, defconfig, yaml, pinmux, doc)
-Usb changes (dts, defconfig, yaml, pinmux, doc)

Last, as every nucleo board, the bard should feature arduino connector.
So please add available arduino connectors (uart and i2c), cf nucleo_f429zi as ref.

File: boards/arm/nucleo_f746zg/Kconfig.defconfig [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+	default "nucleo_f746zg"
+
+if UART_CONSOLE
+
+config UART_STM32_PORT_3
+	default y
+
+endif # UART_CONSOLE
+
+if SPI
```
[the remark this answers is not in the document]
Comment (ajpcode) [reply boards/arm/nucleo_f746zg/Kconfig.defconfig]: I don't believe so, it was a stub I've overlooked removing. Thank you for your detailed feedback, I'll try to address your feedback this evening.

**Comment by ajpcode** [comment]
@erwango After reordering my commits... I've just found the following:
`
arduino_i2c: &i2c1 {};
`
`
arduino_spi: &spi1 {};
`
`
arduino_serial: &usart6 {};
`

... is this what you meant by ... ?
> Last, as every nucleo board, the bard should feature arduino connector.
So please add available arduino connectors (uart and i2c), cf nucleo_f429zi as ref.

**Comment by erwango** [comment]
@ajpcode 

> ... is this what you meant by ... ?

Yes

**Comment by erwango** [comment]
recheck

**Comment by ajpcode** [comment]
@erwango Are the changes I've made acceptable, do you feel I've missed anything?
@dbkinder I believe I've corrected the documentation in line with your feedback - was there anything else you would like to see?

**Review by erwango — APPROVED** [review]

**Comment by erwango** [comment]
@ajpcode , this is fine, thanks! Approved

**Comment by ajpcode** [comment]
Updated documentation in line with @dbkinder 's comment in my other pull request.
Rebased and added SPI support (tested).

**Review by dbkinder — APPROVED** [review]
doc changes LGTM, thanks.

**Review by galak — APPROVED** [review]

**Pull request merged**